lib/Driver/Driver.cpp

```diff
@@ -750,8 +752,21 @@ Driver::buildCompilation(const ToolChain &TC,
 
   if (OFM) {
     if (auto *masterOutputMap = OFM->getOutputMapForSingleOutput()) {
-      C->setCompilationRecordPath(
-          masterOutputMap->lookup(file_types::TY_SwiftDeps));
+      auto buildRecordFilename = masterOutputMap->lookup(file_types::TY_SwiftDeps);
+
+      // In 'emit-module' only mode, suffix build-record filename with
+      // '~moduleonly'. So that module-only mode doesn't mess up build-record
+      // file for full compilation.
+      if (OI.CompilerOutputType == file_types::TY_SwiftModuleFile) {
+        C->setCompilationRecordPath(buildRecordFilename + "~moduleonly");
+      } else {
+        C->setCompilationRecordPath(buildRecordFilename);
+
+        // If we emit module along with full compilation, emit build record
+        // file for '-emit-module' only mode as well.
+        if (OI.ShouldTreatModuleAsTopLevelOutput)
+          C->setOutputCompilationRecordForModuleOnlyBuild();
+      }
```

Comment:
With these changes, this block of code grows so large that I think the clarify would be much improved if it were factored out into a subroutine. That way, it could use an early return to make the flow clearer. The previous version was already a bit unclear to me because it was setting two different things, a record path and a build time, and it is not obvious from the code how they are related. Is there a name for this block that clearly describes its purpose? If not, maybe it should be split up into two or three separate blocks, each with its own if (OFM) test. 
My confusion is not at all your fault, but you have a chance to make things clearer here!